Refactor build system and use pre-built container image for make targets

paulfantom opened the pull request:
Running `go get` in every CI pipeline leads to unpredictable problems due to lack of ability of locking version of downloaded binary. Instead of relying on `go get` we can use pre-built container image with already installed binaries. This is an attempt on using such container. 

Container is build from https://github.com/paulfantom/dockerfiles/blob/master/jsonnet/Dockerfile and stored in [quay.io](https://quay.io/repository/paulfantom/jsonnet-ci)

PR is built on top of #350

/cc @metalmatze @s-urbaniak

paulfantom commented:
I think this is finished. However since with implementing this we will be using pre-built jsonnet binary there is nothing standing in a way of using C++ version supporting `jsonnet fmt`. Additionally we need to distribute container image somehow to be able to use it in CI, with travis this is not a problem (as seen in this PR), but the problem will appear when we will try to port this solution to telemeter repo, which uses only Prow. It seems that DPTP team also jsonnet to validate their dashboards and they are already building an image which could be used in prow (https://github.com/openshift/release/blob/master/projects/origin-release/dashboards-validation/Dockerfile). Before merging it I will try to figure out if we can modify that image and use it in those scenarios.

/hold

squat commented:
@paulfantom we could also lock the version by doing, e.g. `go get github.com/jsonnet-bundler/jsonnet-bundler/cmd/jb@v2.0.0`. I've had a lot of success doing this in some projects.

paulfantom commented:
@squat that was my first try. Sadly we cannot do this in case of go-jsonnet as it is not supported.

```
$ docker run -it golang:1.11 bash
root@a26f4c2e2d09:/go# go get github.com/google/go-jsonnet/cmd/jsonnet@v0.12.1        
go: cannot use path@version syntax in GOPATH mode
root@a26f4c2e2d09:/go# go get github.com/google/go-jsonnet/jsonnet@v0.12.1
go: cannot use path@version syntax in GOPATH mode
```

There is some discussion about that specific error in go repo: https://github.com/golang/go/issues/29415

brancz commented:
I'd be happy with doing this, it could also allow us to speed up CI as building that c++ based version is pretty intense.

paulfantom commented:
Added refactored Makefile in the same way as in https://github.com/coreos/prometheus-operator/pull/2604 and changed jsonnet image to the one used in prometheus-operator.

brancz commented:
/lgtm

just waiting for ART to confirm we can go ahead with 1.12 for 4.2

brancz commented:
/hold

brancz commented:
feel free to remove hold if confirmed

brancz commented on .travis.yml:
```diff
@@ -1,14 +1,23 @@
+go_import_path: github.com/openshift/cluster-monitoring/operator
+#env:
+#  - GO111MODULE=on
+#cache:
```
Are you trying something out here, or is this intentionally commented out?

paulfantom replied:
trying out ;)

paulfantom commented:
Blocked by https://github.com/coreos/prometheus-operator/pull/2609

paulfantom commented:
It is ready now.

/hold cancel

/cc: @brancz @s-urbaniak @squat

squat commented:
/test e2e-aws-upgrade
